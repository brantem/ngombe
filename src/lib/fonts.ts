import { Nunito } from 'next/font/google';

export const nunito = Nunito({ subsets: ['latin'], weight: ['500', '800'] });
