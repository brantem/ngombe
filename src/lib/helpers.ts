import { twMerge } from 'tailwind-merge';
import clsx, { ClassValue } from 'clsx';

/* c8 ignore next 3 */
export const cn = (...inputs: ClassValue[]) => {
  return twMerge(clsx(inputs));
};

export const calcPercentage = (a: number, b: number) => {
  return Math.round((a / b) * 100);
};
