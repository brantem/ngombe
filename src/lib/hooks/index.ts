import { useState, useLayoutEffect, useEffect } from 'react';

const useIsomorphicLayoutEffect = typeof window !== 'undefined' ? useLayoutEffect : useEffect;

export const useCurrentHeight = () => {
  const [height, setHeight] = useState(0);

  useIsomorphicLayoutEffect(() => {
    const handleResize = () => {
      if (typeof window === 'undefined') return;
      setHeight(window.innerHeight);
    };

    handleResize();
    window.addEventListener('resize', handleResize);

    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return height;
};
