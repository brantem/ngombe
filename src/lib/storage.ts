import { openDB, IDBPDatabase, StoreNames, StoreValue, StoreKey } from 'idb';

import { Schema } from 'types/storage';

/* c8 ignore start */
class Storage {
  db: IDBPDatabase<Schema> | undefined;

  _getDB() {
    return openDB('ngombe', 1, {
      async upgrade(db) {
        db.createObjectStore('goals', { keyPath: 'timestamp' });

        db.createObjectStore('records', { keyPath: 'timestamp' });
      },
    });
  }

  async get<Name extends StoreNames<Schema>>(name: Name, key: StoreKey<Schema, Name>) {
    const db = await this._getDB();
    return db.get(name, key);
  }

  async getAll<Name extends StoreNames<Schema>>(name: Name, query: IDBKeyRange) {
    const db = await this._getDB();
    return db.getAll(name, query);
  }

  async put<Name extends StoreNames<Schema>>(name: Name, value: StoreValue<Schema, Name>) {
    const db = await this._getDB();
    return db.put(name, value);
  }

  async add<Name extends StoreNames<Schema>>(name: Name, value: StoreValue<Schema, Name>) {
    const db = await this._getDB();
    return db.add(name, value);
  }

  async delete<Name extends StoreNames<Schema>>(name: Name, key: StoreKey<Schema, Name>) {
    const db = await this._getDB();
    return db.delete(name, key);
  }
}

const storage = new Storage();
export default storage;
/* c8 ignore stop */
