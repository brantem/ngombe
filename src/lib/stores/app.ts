import { create } from 'zustand';

import { goalStore, recordsStore } from 'lib/stores';

type AppState = {
  items: Record<string, any>;
  setItem: (key: string, value: any) => void;
  deleteItem: (key: string) => void;

  date: string | undefined; // YYYY-MM-DD
  setDate(date: string | undefined): void;
};

export const useAppStore = create<AppState>()((set) => ({
  items: {},
  setItem(key, value) {
    set((state) => ({ items: { ...state.items, [key]: value } }));
  },
  deleteItem(key) {
    set((state) => {
      const items = state.items;
      delete items[key];
      return { items };
    });
  },

  date: undefined,
  setDate(date) {
    set({ date });
    goalStore.getState().fetch(date);
    recordsStore.getState().fetch(date);
  },
}));
