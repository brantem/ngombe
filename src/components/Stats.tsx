import { useGoalStore, useRecordsStore } from 'lib/stores';
import { calcPercentage, cn } from 'lib/helpers';

type StatsProps = React.ComponentPropsWithoutRef<'div'> & {
  isBackground?: boolean;
};

const Stats = ({ className, isBackground, ...props }: StatsProps) => {
  const goal = useGoalStore((state) => state.value);
  const value = useRecordsStore((state) => state.calcTotalValue());
  const percentage = calcPercentage(value, goal);

  return (
    <div className={cn('text-center flex flex-col justify-center items-center', className)} {...props}>
      <h1
        className={cn(
          'tabular-nums text-7xl font-extrabold',
          /* c8 ignore next */ isBackground ? 'text-neutral-200' : 'text-cyan-600',
        )}
        data-testid="stats-value"
      >
        {value}ml
      </h1>
      <p
        className={cn(
          'tabular-nums text-2xl',
          /* c8 ignore next */ isBackground ? 'text-neutral-300' : 'text-cyan-500',
        )}
        data-testid="stats-percentage"
      >
        {percentage}% of your goal
      </p>
    </div>
  );
};

export default Stats;
