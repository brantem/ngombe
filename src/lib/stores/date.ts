import { create } from 'zustand';

import { goalStore, recordsStore } from 'lib/stores';

type DateState = {
  value: string | undefined;
  set(value: string | undefined): void;
};

// it didn't work when I used useState for this
export const useDateStore = create<DateState>()((set) => ({
  value: undefined,
  set(value) {
    set({ value });
    goalStore.getState().fetch(value);
    recordsStore.getState().fetch(value);
  },
}));
