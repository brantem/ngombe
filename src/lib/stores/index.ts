import { useEffect } from 'react';

import { recordsStore } from 'lib/stores';

/* c8 ignore start */
export const useLoadStore = () => {
  useEffect(() => {
    (async () => {
      recordsStore.getState().setDate(undefined);
    })();
  }, []);
};
/* c8 ignore stop */

export * from 'lib/stores/modal';
export * from 'lib/stores/goal';
export * from 'lib/stores/records';
