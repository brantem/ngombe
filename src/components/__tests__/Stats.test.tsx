import { renderHook, act, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import Stats from 'components/Stats';

import { goalStore, useHistoriesStore } from 'lib/stores';

describe('Stats', () => {
  it('should not show NaN when not ready', () => {
    render(<Stats />);
    expect(screen.getByTestId('stats-value').textContent).toEqual('0ml');
    expect(screen.getByTestId('stats-percentage').textContent).toEqual('0% of your goal');
  });

  it('should show correct stats', () => {
    act(() => goalStore.setState({ goal: 2500, isReady: true }));
    const histories = renderHook(() => useHistoriesStore());

    const { container } = render(<Stats />);
    act(() => histories.result.current.drink(Date.now().toString(), 100));
    expect(screen.getByTestId('stats-value').textContent).toEqual('100ml');
    expect(screen.getByTestId('stats-percentage').textContent).toEqual('4% of your goal');
    expect(container).toMatchSnapshot();
  });
});
