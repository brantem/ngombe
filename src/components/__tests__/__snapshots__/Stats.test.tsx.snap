// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Stats > should show correct stats 1`] = `
<div>
  <div
    class="text-center flex flex-col justify-center items-center"
  >
    <h1
      class="tabular-nums text-7xl font-extrabold text-cyan-600"
      data-testid="stats-value"
    >
      100
      ml
    </h1>
    <p
      class="tabular-nums text-2xl text-cyan-500"
      data-testid="stats-percentage"
    >
      4
      % of your goal
    </p>
  </div>
</div>
`;
